search/custom-embeddings/custom_embeddings.ipynb

```diff
@@ -0,0 +1,800 @@
+    "\n",
+    "\n",
+    "def scrape_question(question_url: str) -> str:\n",
+    "    response = requests.get(question_url)\n",
+    "\n",
+    "    if response.status_code != 200 or not response.content:\n",
+    "        print(f\"URL: {question_url} Code: {response.status_code}\")\n",
+    "        return None\n",
+    "\n",
+    "    print(f\"Scraping {question_url}\")\n",
+    "\n",
+    "    link_title = response.url.split(\"/\")[-1] + \".html\"\n",
+    "    gcs_uri = f\"{GCS_URI_PREFIX}{link_title}\"\n",
+    "\n",
+    "    # Upload HTML to Google Cloud Storage\n",
+    "    blob = bucket.blob(f\"{BLOB_PREFIX}{link_title}\")\n",
+    "    blob.upload_from_string(response.content, content_type=HTML_MIME_TYPE)\n",
+    "    time.sleep(1)\n",
+    "    return gcs_uri"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Get the published URL from the ID\n",
+    "QUESTION_BASE_URL = \"https://stackoverflow.com/questions/\"\n",
+    "df[\"question_url\"] = df[\"id\"].apply(lambda x: f\"{QUESTION_BASE_URL}{x}\")\n",
+    "\n",
+    "# Scrape HTML from stackoverflow.com and upload to GCS\n",
+    "df[\"gcs_uri\"] = df[\"question_url\"].apply(scrape_question)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Restructure the embeddings data to follow Vertex AI Search format (Unstructured with Metadata)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "EMBEDDINGS_FIELD_NAME = \"embedding_vector\"\n",
+    "\n",
+    "\n",
+    "def format_row(row):\n",
+    "    return {\n",
+    "        \"id\": row[\"id\"],\n",
+    "        \"content\": {\"mimeType\": HTML_MIME_TYPE, \"uri\": row[\"gcs_uri\"]},\n",
+    "        \"structData\": {\n",
+    "            {EMBEDDINGS_FIELD_NAME}: row[\"embedding\"],\n",
+    "            \"title\": row[\"title\"],\n",
+    "            \"body\": row[\"body\"],\n",
+    "            \"answer_count\": row[\"answer_count\"],\n",
+    "            \"question_url\": row[\"question_url\"],\n",
+    "        },\n",
+    "    }\n",
+    "\n",
+    "\n",
+    "jsonl_filename = \"vais_embeddings.jsonl\"\n",
+    "\n",
+    "with open(jsonl_filename, \"w\") as f:\n",
+    "    f.write(\n",
+    "        df.apply(format_row, axis=1)\n",
+    "        .to_json(orient=\"records\", lines=True, force_ascii=False)\n",
+    "        .replace(\"\\/\", \"/\")  # To prevent escaping the / characters\n",
+    "    )"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Upload the JSONL file to Google Cloud Storage"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "embeddings_file = f\"gs://{BUCKET_NAME}/{DIRECTORY}/{jsonl_filename}\"\n",
+    "!gsutil mv {jsonl_filename} {embeddings_file}"
```

Comment:
At the top of the notebook you installed the `google-cloud-storage` library. Should you switch to using the Python client here to avoid users having to make sure that Python API auth *and* CLI auth is handled?

Reply:
Probably a good idea, this was quicker when rapidly prototyping